File: src/Binary_Tree_Inorder_Traversal.py
# -*- coding: utf-8 -*-


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    # @param {TreeNode} root
    # @return {integer[]}
    def inorderTraversal(self, root):
        ret, st = [], []
        node = root
        while node or len(st) != 0:
            if node:
                st.append(node)
                node = node.left
            else:
                node = st.pop()
                ret.append(node.val)
                node = node.right
        return ret


if __name__ == "__main__":
    solution = Solution()
    a, b = TreeNode(1), TreeNode(2)
    c = TreeNode(3)
    a.right, b.left = b, c
    print solution.inorderTraversal(a)

File: src/algorithms/cpp/Binary_Tree_Inorder_Traversal.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
#include <stack>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector <int> inorderTraversal(TreeNode* root) {
        TreeNode *node = root;
        stack <TreeNode*> st;
        vector <int> result;
        while (node || !st.empty()) {
            if (node) {
                st.push(node);
                node = node->left;
            }
            else {
                node = st.top();
                st.pop();
                result.push_back(node->val);
                node = node->right;
            }
        }
        return result;
    }
};

int main() {
    return 0;
}
